Update repository to current elm version

andys8 opened the pull request:
Hi @soenkehahn, I made a PR to update this example repository because it was using Elm 0.17 which is not the current version anymore. There we're breaking changes in Elm and I migrated the code. I adapted the tests and changed the structure they should have. The dependencies of stack are updated in order to have the updated haskell to elm dependencies. I made minor adaptions to the haskell interface and the elm code generation to get it working again.

Starting the server using `stack exec server` and adding and deleting items works as expected.

### TL;DR
* Updated stack dependencies
* Updated elm to 0.18
* Elm code migration
* Haskell code generation fixed
* Fixes #3 

### Open issue

The haskell tests aren't working :( I'm not sure if this is due to the changes or the update of dependencies. The error message looks pretty similar to the one posted in a different repositories issue: https://github.com/haskell-servant/example-servant-minimal/issues/5 . Perhaps you can help me out here. 

```sh
example-servant-elm/server/test/AppSpec.hs:20:1: error:
    • Couldn't match expected type ‘Manager -> BaseUrl -> ClientM [ItemId]’
                  with actual type ‘ClientM [ItemId]’
    • When checking that the inferred type
        getItemIds :: ClientM [ItemId]
      is as general as its signature
        getItemIds :: Manager -> BaseUrl -> ClientM [ItemId]
```

---

By the way: There is another repository of https://github.com/yannick-cw/paraglidingin_front_elm where the author @yannick-cw set up an servant elm example application.

Happy to hear feedback :)

soenkehahn commented:
Awesome that you're looking into this.

I pushed this commit that fixes the server tests: https://github.com/haskell-servant/example-servant-elm/commit/d9614233e364a1a4cda66b86225317a1339c798b. Feel free to cherry-pick that into your branch.

Note that this repo uses `Makefile`s to run commands (and document those commands). Running the client test-suite (with `cd client ; make run-tests`) doesn't work anymore -- I think because you restructured the tests? Can you fix that, please?

soenkehahn commented:
Btw, I realized we didn't have CI for this repo, so I set that up. You should rebase your branch on current master, then we'll get automatic checks for your PR.

andys8 commented:
Thanks a lot. I fixed elm-test and the integration. The make file compiles elm and elm-test does it, too. This takes a long time on travis (known issue: https://github.com/elm-lang/elm-compiler/issues/1473 and https://github.com/elm-community/elm-test/issues/42). `elm-test` has to be installed globally and I added it to the `README.md`.

andys8 commented on server/src/Api.hs:
```diff
@@ -9,21 +9,21 @@ import           Data.Aeson
 -- types
 
-type ItemId = Integer
+type ItemId = Int
```
With `type ItemId = Integer` there is this error:
```sh
Preprocessing executable 'server' for example-servant-elm-0.0.0...
[1 of 3] Compiling Api              ( server/src/Api.hs, .stack-work/dist/x86_64-osx/Cabal-1.24.2.0/build/server/server-tmp/Api.o )

/Users/ascharf/dev/repository/example-servant-elm/server/src/Api.hs:35:10: error:
    • No instance for (ElmType Integer)
        arising from a use of ‘elm-export-0.6.0.1:Elm.Type.$dmtoElmType’
    • In the expression: elm-export-0.6.0.1:Elm.Type.$dmtoElmType @Item
      In an equation for ‘toElmType’:
          toElmType = elm-export-0.6.0.1:Elm.Type.$dmtoElmType @Item
      In the instance declaration for ‘ElmType Item’
Progress: 1/2
--  While building package example-servant-elm-0.0.0 using:
      /Users/ascharf/.stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_1.24.2.0_ghc-8.0.2 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.24.2.0 build exe:server test:spec --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
make: *** [slow-test] Error 1
```

I'm not sure what's causing it, but changing it to `Int` solved it.

soenkehahn replied:
Ah, I see. It seems that `ElmType` doesn't support `Integer`. That's fine then.

andys8 commented on stack.yaml:
```diff
@@ -1,16 +1,8 @@
-      git: https://www.github.com/mattjbray/servant-elm
-      commit: 28cb98a
-    extra-dep: True
+    extra-dep: true
```
kk

andys8 commented on client/Main.elm:
```diff
@@ -59,9 +56,10 @@ type Msg
-    | Delete ItemId
+    = Initial (Result Http.Error (List ItemId))
+    | NewItem (Result Http.Error Item)
+    | CreatedItem (Result Http.Error ItemId)
```
Exactly. This was the major braking change where `Http.send` want's a type constructor as argument and the user is handling a result. There could be ways to reduce boilerplate or clutter if all error's should be handled the same way. The way it's implemented now is the most straight forward way. In the [guide](https://guide.elm-lang.org/architecture/effects/http.html) it's done the same using pattern matching instead of `case-of`.

andys8 commented on README.md:
```diff
@@ -13,7 +13,7 @@ This example project illustrates how to set up a project that:
 There's a `Makefile` included with the following targets:
 
 - `setup` -- Set up everything: install ghc and dependencies. (Needs `stack` and
-  `elm`.)
```
Changed it.

andys8 commented:
@soenkehahn Thanks for looking into it and the great feedback. I made some adjustments to the commit and replied. The changes we're necessary to have the project running, compiling and being tested again. Looking over the changes I wouldn't say there are things that could be done in another PR. Testing is as it is done now is how it's done in most of the projects. I'm not sure how it worked before, but I wanted the example to have a working test setup again.

soenkehahn commented:
Sorry, I should have been clearer. I think switching to `elm-test` is perfectly fine, since -- as you said -- that's the more common way to run tests in elm. Actually I think that's a great change!

I was mostly talking about the addition of `CreatedItem` and the wrapping of all the fields of the `FromServer` constructors in `Result Http.Error`. Before the error handling was very separated from the rest of the code, which I thought was nice. And I suspect that we can do the same in the new version. I'll try to come up with a commit on top of your branch, but I'm not sure when I'll get to that.

One note, just in case it's the cause of some confusion: `Http.send` does not necessarily need a constructor as its first argument. Just a function of type `(Result Error a -> msg)`. That *can* be a constructor -- constructors are functions with the field type as arguments and the constructed type as their output type -- but it could be any function that constructs an `msg` from a `Result Error a`. So it could be a function that pattern-matched on the `Result` and created an `Msg`. (Please excuse my lecturing if you know all this.)

andys8 commented:
Okay, you're detailed explanation helped a lot. I think we can extract the handling in a similar way it was done before.

I made the changes I'd expected you would do them in a further commit:
https://github.com/haskell-servant/example-servant-elm/pull/5/commits/592a2db83f2e7c1aaca905a029877c92f71eecb0?diff=split

Let me know what you think and feel free to commit changes.

soenkehahn commented:
Hi @andys8. Sorry for the long wait.

Yes, those changes were exactly what I was talking about regarding error handling.

I've added two commits on top of your branch here: https://github.com/haskell-servant/example-servant-elm/tree/version-updates

I'm fairly sure that the first commit (https://github.com/haskell-servant/example-servant-elm/commit/53fcc7ed33853a55096c6a8f142eba861b5cf7a6) is a change we want. It removes one more roundtrip to the server and behaves the same apart from that.

The second commit (https://github.com/haskell-servant/example-servant-elm/commit/99333c365493a950427c0afd35af8b0d9dfde8dd) reverts a change to the delete endpoint of the rest api (while still keeping the functionality intact). It needs a workaround in the generated elm code (adding the `NoContent` type). So I'm less opinionated about this commit, since it's a trade-off in that sense. I would prefer a cleaner rest api and accept the slightly weird workaround in the code generation for that. Let me know what you think.

(Generally this type should be provided by `servant-elm`'s `defElmImports`, since it's a type that `servant` provides, I think. But one step at a time... :) )

I'll give you push access to this repo. Please feel free to push these changes to master.

andys8 commented:
No problem. The changes you made are looking very good to me. One small improvement: https://github.com/haskell-servant/example-servant-elm/commit/fd8621eed10693e1015dab1a150f4cfba16c4303. What do you think of ignoring the response content to get rid of the workaround to add the `NoContent` type manually as string?

soenkehahn commented:
I don't think that works if you actually run the code generation script (through e.g. `make slow-test`). Because the generated elm code will still expect the type `NoContent` to exist.

andys8 commented:
You're right. Didn't think about the generation. Removed the commit from the branch.
https://github.com/haskell-servant/example-servant-elm/commits/version-updates

Therefore I'm good with it.

soenkehahn commented:
Alright, I merged it. Thanks a lot for working on this and for your patience.
